TL-9 Implement multiple page Todo app

BiroAron opened the pull request.

BiroAron committed:
ITL 33 - Add a Login page

BiroAron committed:
ILT 34 - Add a Sign Up page

BiroAron committed:
ITL 35 - Create routes

BiroAron committed:
ITL 9 - Made some changes on the frontend

- made changes to the frontend so it uses the operations on the backend

BiroAron committed:
ITL 9 - Restructured the backend

BiroAron committed:
ITL 9 - Made a reusable input field and used it

SzaszWebgurus commented on client/src/api.ts:
```diff
@@ -0,0 +1,132 @@
+import axios from 'axios'
+
+const API_BASE_URL = 'http://localhost:8080'
```
use .env

SzaszWebgurus commented on client/src/api.ts:
```diff
@@ -0,0 +1,132 @@
+      throw new Error('JWT token not found')
+    }
+
+    const response = await axios.get(`${API_BASE_URL}/users/get-all-todos`, {
```
GET /todos

SzaszWebgurus commented on client/src/views/Dashboard.vue:
```diff
@@ -150,26 +157,33 @@ function setSearchQuery(searchSentence: string) {
+async function addTodo(todo: Todo) {
+  try {
+    await addNewTodo(
+      todo.title,
```
why don't you pass as an object?

SzaszWebgurus commented on client/src/views/Dashboard.vue:
```diff
@@ -150,26 +157,33 @@ function setSearchQuery(searchSentence: string) {
+      todo.isChecked,
+      todo.date
+    )
+    fetchAndAddTodos()
```
you don't need to fetch each todo after the add, you just need to append to the list

SzaszWebgurus commented on client/src/views/Dashboard.vue:
```diff
@@ -194,16 +208,21 @@ function setSortOrderDescending() {
+  try {
+    //console.log('TodoId: ' + todo.title)
+    await editCurrentTodo(
+      todo._id,
```
same as above

SzaszWebgurus commented on client/src/views/Dashboard.vue:
```diff
@@ -194,16 +208,21 @@ function setSortOrderDescending() {
+      todo.isChecked,
+      todo.date
+    )
+    fetchAndAddTodos()
```
same

SzaszWebgurus commented on client/src/views/Dashboard.vue:
```diff
@@ -216,12 +235,17 @@ function updateItemList(index: number) {
-function saveToLocalStorage() {
-  localStorage.setItem('todos', JSON.stringify(todos))
-}
+// function saveToLocalStorage() {
```
remove

SzaszWebgurus commented on client/src/views/Login.vue:
```diff
@@ -0,0 +1,81 @@
+
+<script setup lang="ts">
+import { RouterLink, useRouter } from 'vue-router'
+import { loginUser } from '../api'
```
use better naming here insted of `api`

SzaszWebgurus commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,125 @@
+<template>
+  <body class="flex h-screen w-screen bg-white">
+    <div
+      class="m-auto w-full max-w-lg rounded-2xl border-4 border-black bg-[#fff7f7] p-5 shadow-xl"
```
define the color in the tailwind config

SzaszWebgurus commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,125 @@
+        </div>
+      </header>
+      <form @submit.prevent="register">
+        <div class="mb-2 flex flex-row space-x-2">
```
flex-row is the default

SzaszWebgurus commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,125 @@
+async function register() {
+  try {
+    await registerUser(
+      email.value,
```
pass object

SzaszWebgurus commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  const token = await loginUser(email.value, password.value)
+  console.log('JWT Token:', token)
```
do you need this?

SzaszWebgurus commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,125 @@
+          :input-value="password"
+          input-name="Password"
+          input-type="password"
+          @update-value="updatePassword"
```
pass the to be changed variable as a v-model

SzaszWebgurus commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,125 @@
+  router.push({ name: 'Dashboard' })
+}
+
+function updateEmail(newEmail: string) {
```
you won't need these if you pass the variable as v-model

SzaszWebgurus commented on server/src/index.ts:
```diff
@@ -0,0 +1,43 @@
+app.use(bodyParser.json());
+
+const MONGO_URL =
+  "mongodb+srv://biroaron:123@cluster0.y7nbxkk.mongodb.net/ToDo?retryWrites=true&w=majority";
```
use .env

SzaszWebgurus commented on server/src/index.ts:
```diff
@@ -0,0 +1,43 @@
+
+mongoose.Promise = Promise;
+
+// const connectOptions: mongoose.ConnectOptions = {
```
🐒

SzaszWebgurus commented on server/src/index.ts:
```diff
@@ -0,0 +1,43 @@
+
+app.use("/", router);
+
+const PORT = 8080;
```
.env

SzaszWebgurus commented on server/src/index.ts:
```diff
@@ -0,0 +1,43 @@
+const PORT = 8080;
+
+app.listen(PORT, () => {
+  console.log(`Server running on http://localhost:${PORT}/`);
```
.env for the host

SzaszWebgurus commented on server/src/models/todo.ts:
```diff
@@ -0,0 +1,16 @@
+  },
+  description: String,
+  isChecked: { type: Boolean, default: false },
+  date: { type: Date, default: new Date("2023-08-02") },
```
hardcoded date?

SzaszWebgurus commented on server/src/models/todo.ts:
```diff
@@ -0,0 +1,16 @@
+  title: { type: String, required: true },
+  priority: {
+    type: String,
+    enum: ["High", "Medium", "Low"],
```
move this to a separate exported enum

SzaszWebgurus commented on server/src/models/todo.ts:
```diff
@@ -0,0 +1,16 @@
+  priority: {
+    type: String,
+    enum: ["High", "Medium", "Low"],
+    default: "Medium",
```
here you can use the value of the exported enum from above

SzaszWebgurus commented on server/src/models/todo.ts:
```diff
@@ -0,0 +1,16 @@
+  date: { type: Date, default: new Date("2023-08-02") },
+});
+
+export default mongoose.model("Todo", TodoSchema);
```
lowercase for the `todo`

SzaszWebgurus commented on server/src/models/user.ts:
```diff
@@ -0,0 +1,14 @@
+  },
+});
+
+export default mongoose.model("User", UserSchema);
```
lowercase

SzaszWebgurus commented on server/src/router/todos.ts:
```diff
@@ -0,0 +1,17 @@
+  deleteTodo,
+  addTodo,
+} from "../controller/todos";
+//import { extractUserId } from "../middleware/index";
```
🐘

SzaszWebgurus commented on server/src/router/todos.ts:
```diff
@@ -0,0 +1,17 @@
+
+const router = Router();
+
+router.post("/add-todo", addTodo);
```
POST `/` the todo comes from the context

SzaszWebgurus commented on server/src/router/todos.ts:
```diff
@@ -0,0 +1,17 @@
+const router = Router();
+
+router.post("/add-todo", addTodo);
+router.get("/get-all-todos", getAllTodos);
```
same as above, apply for each endpoint pls

SzaszWebgurus commented on server/src/router/users.ts:
```diff
@@ -0,0 +1,21 @@
+  getUser,
+} from "../controller/users";
+import { isAuthenticated } from "../middleware/index";
+//import { extractUserId } from "../middleware/index";
```
🐸

SzaszWebgurus commented on server/src/router/users.ts:
```diff
@@ -0,0 +1,21 @@
+  updateUser,
+  getUser,
+} from "../controller/users";
+import { isAuthenticated } from "../middleware/index";
```
better naming than `index`

tamasm-wg commented on client/package.json:
```diff
@@ -10,10 +10,12 @@
   },
   "dependencies": {
     "@vueuse/core": "^10.3.0",
+    "axios": "^1.4.0",
```
I would suggest using fetch() instead of axios, since that is already included, and is more than enough for the kind of requests you are making. 
More here : https://blog.logrocket.com/axios-vs-fetch-best-http-requests/

tamasm-wg commented on client/src/views/Dashboard.vue:
```diff
@@ -150,26 +157,33 @@ function setSearchQuery(searchSentence: string) {
+async function deleteItem(_id) {
+  try {
+    await deleteCurrentTodo(_id)
+    fetchAndAddTodos()
```
Here also just remove the current todo from the list instead of fetching every time you delete one

BiroAron replied:
I did this already in my latest commits

tamasm-wg commented on client/src/views/Dashboard.vue:
```diff
@@ -194,16 +208,21 @@ function setSortOrderDescending() {
+    )
+    fetchAndAddTodos()
+  } catch (error) {
+    console.error('Error:', error)
```
this is just a suggestion : when you logging errors it is useful to add the function name where the error occurred, so it easier to track down errors ( for ex. console.error('[editCurrentTodo Error]', error)

BiroAron committed:
ITL 9 - Made changes regarding the comments

- removed commented lines an unnecesary consol.log-s
- used v-model instead of passing props into Baseinput.vue
- replaced fetchAllTodos function where it was possible

BiroAron committed:
ITL 9 - Made some additional changes

- removed unnecessary functions from signup and login pages

BiroAron committed:
ITL 9 - Made the last change

- used TodoPriorities enum to give a default value to the priority of a todo

SzaszWebgurus commented on client/src/router.ts:
```diff
@@ -0,0 +1,17 @@
+import Dashboard from './views/Dashboard.vue'
+
+const routes = [
+  { path: '', name: 'Login', component: Login },
```
path: '/login', and if the user is logged in add a redirect to the dashboard, otherwise to the login

SzaszWebgurus commented on client/src/views/Login.vue:
```diff
@@ -0,0 +1,71 @@
+<template>
+  <body class="flex h-screen w-screen bg-white">
+    <div
+      class="m-auto w-full max-w-sm rounded-2xl border-4 border-black bg-[#fff7f7] p-5 shadow-xl"
```
move the color to the tailwind config

BiroAron committed:
ITL 9 - Made changes regarding the comments

- added a redirect to the router
- added custom color to tailwind config
- in addition also fixed a bug regarding the display of the date

BiroAron committed:
ITL 9 - Created a logout button

tamasm-wg commented on server/.env:
```diff
@@ -1 +1,3 @@
-MONGODB_URI="mongodb+srv://emanueldrusu:Mongowebgurus22@wgclusterlearn.gr0pdlo.mongodb.net/?retryWrites=true&w=majority"
\ No newline at end of file
+MONGO_URL = "mongodb+srv://biroaron:123@cluster0.y7nbxkk.mongodb.net/ToDo?retryWrites=true&w=majority"
```
A better approach here is to have different variables for user and password like :
MONGO_USER=
MONGO_PASSWORD=
Then you build your connection string in the file where you use it

tamasm-wg commented on server/src/index.ts:
```diff
@@ -0,0 +1,35 @@
+mongoose.Promise = Promise;
+
+mongoose
+  .connect(process.env.MONGO_URL)
```
you can pass the connection string without the query parameters ( ["mongodb+srv://emanueldrusu:[Mongowebgurus22@wgclusterlearn.gr0pdlo.mongodb.net](mailto:Mongowebgurus22@wgclusterlearn.gr0pdlo.mongodb.net)](mongodb+srv://biroaron:123@cluster0.y7nbxkk.mongodb.net/)" ) and add those here in the options :
.connect(connectionString, { retryWrites: true, w: 'majority', dbName: '...' })

tamasm-wg commented on server/src/router/todos.ts:
```diff
@@ -0,0 +1,18 @@
+} from "../controller/todos";
+
+const router = Router();
+
```
These routes should also be protected by the auth middleware

BiroAron replied:
These are protected because the full link is localhost:8080/user/todo and im using isAuthenticated middleware for every user related query. You can check it in "server/src/router/users.ts"

tamasm-wg commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,98 @@
+      <footer>
+        <RouterLink
+          class="float-left text-sm text-black hover:text-low"
+          href="#"
```
remove href

tamasm-wg commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,98 @@
+        >
+        <RouterLink
+          class="float-right text-sm text-black hover:text-low"
+          href="#"
```
remove href

tamasm-wg commented on client/src/views/SignUp.vue:
```diff
@@ -0,0 +1,98 @@
+async function register() {
+  try {
+    await registerUser(user)
+  } catch (error) {
```
You should implement showing the user that the registration failed, not only logging it to the console.

tamasm-wg commented on client/src/views/Login.vue:
```diff
@@ -0,0 +1,71 @@
+  try {
+    await loginUser(email.value, password.value)
+    router.push({ name: 'Dashboard' })
+  } catch (error) {
```
Display some feedback to the user if login failed , not just in the console

BiroAron committed:
ITL 9 - Made all the changes requested

BiroAron committed:
ITL 9 - Added full routing on server side

tamasm-wg commented on client/src/components/inputfields/BaseInput.vue:
```diff
@@ -0,0 +1,31 @@
+  <div class="mb-2 flex flex-col">
+    <label class="mb-1 block text-black">{{ inputName }}</label>
+    <input
+      v-model="email"
```
since this is a generic input field , you should use a more generic variable name for the v-model

BiroAron committed:
ITL 9 - Made some minor fixes

tamasm-wg commented on client/src/views/Dashboard.vue:
```diff
@@ -0,0 +1,202 @@
+  toggleNewTodo()
+}
+
+function deleteNewItem() {
```
if this function only does that calls another function , you can call it directly instead

tamasm-wg commented on client/src/views/Dashboard.vue:
```diff
@@ -0,0 +1,202 @@
+}
+function updateItemList(index: number) {
+  const todoSave: Todo = todos[index]
+  todos.splice(index, 1)[0]
```
what this line does? you splice the todos and grab the first element but you didn't assign it to any variable

tamasm-wg commented on client/src/views/Dashboard.vue:
```diff
@@ -0,0 +1,202 @@
+}
+
+async function fetchAndAddTodos() {
+  console.log('fetchAndAddTodos')
```
remove this console.log please

SzaszWebgurus commented on server/src/index.ts:
```diff
@@ -0,0 +1,44 @@
+    `${process.env.MONGO_BASE_URL}://${process.env.MONGO_USER}:${process.env.MONGO_PASSWORD}@${process.env.MONGO_CLUSTER}.${process.env.MONGO_HOSTNAME}/`,
+    connectOptions
+  )
+  .then(() => {
```
use async function instead of callback

SzaszWebgurus commented on server/src/models/user.ts:
```diff
@@ -0,0 +1,13 @@
+  firstname: { type: String, required: true },
+  lastname: { type: String, required: true },
+  email: { type: String, required: true },
+  authentication: {
```
append these into a single password string (you'll know the salt length, so you can easily split them based on the length)

BiroAron committed:
ITL 9 - Made all the changes requested on the pr
